tools/clang_tools/README.md

```diff
@@ -0,0 +1,60 @@
+# Envoy Clang libtool developer tools
+
+## Overview
+
+A number of tools live in this directory that are intended for use by Envoy
+developers (and potentially CI). These are host tools and should not be linked
+into the Envoy target. They are based around Clang's
+[libtooling](https://clang.llvm.org/docs/LibTooling.html) libraries, a C++
+framework for writing Clang tools in the style of `clang-format` and
+`clang-check`.
+
+## Building and running
+
+To build tools in this tree, a Clang binary install must be available. If you
+are building Envoy with `clang`, this should already be true of your system. You
+can find prebuilt binary releases of Clang at https://releases.llvm.org. You
+will need the Clang version used by Envoy in CI (currently clang-9.0).
+
+To build a tool, set the following environment variables:
+
+```console
+export CLANG_TOOLS_LIB=<path to clang installation>/lib
+export CLANG_TOOLS_INCLUDE=<path to clang installation>/include
+```
+
+Assuming that `CC` and `CXX` already point at Clang, you should be able to build
+with:
+
+```console
+bazel build //tools/clang_tools/syntax_only
+```
+
+To run `libtooling` based tools against Envoy, you will need to first generate a
+compilation database, which tells the tool how to take a source file and locate
+its various dependencies. The `tools/gen_compilation_database.py` script
+generates this and also does setup of the Bazel cache paths to allow external
+dependencies to be located:
+
+```console
+tools/gen_compilation_database.py --run_bazel_build --include_headers
+```
+
+Finally, the tool can be run against source files in the Envoy tree. Due to the
+nature of the treatment of Clang's [builtin
+headers](https://clang.llvm.org/docs/LibTooling.html#builtin-includes), it's
+necessary to point the tool at these as well.
+
+```console
+bazel-bin/tools/clang_tools/syntax_only/syntax_only \
+  --extra-arg="-isystem<path to clang installation>/lib/clang/9.0.0/include/"
```

Comment:
Interesting, this shouldn't be needed at runtime but at compilation generating, do you know which headers will be missing without this?

Reply:
Things like `stddef.h`, see the linked explanation of how `libtooling` finds these headers; keep in mind that the "runtime" here is basically "compile time", i.e. it's running the Clang AST frontend and so needs to find headers belonging to the source under analysis (via transitive `#include`).

Reply:
yes I know the "runtime" here is compile time, but I think clang does implicitly add that include path. Maybe because the compilation db generator force update clang path to `clang++`? I think we can remove that since we no longer have cc_wrapper.

Reply:
We could try not forcing to `clang++`, but from the above link:

Clang tools need their builtin headers and search for them the same way Clang does. Thus, the default location to look for builtin headers is in a path $(dirname /path/to/tool)/../lib/clang/3.3/include relative to the tool binary. This works out-of-the-box for tools running from llvm’s toplevel binary directory after building clang-resource-headers, or if the tool is running from the binary directory of a clang install next to the clang binary.

Tips: if your tool fails to find stddef.h or similar headers, call the tool with -v and look at the search paths it looks through.

Reply:
I think this was magically fixed as a result of the path changes, so resolving.